### ISSUE TITLE
Using with Gitlab private registry

### ISSUE THREAD

**Issue by panzelva** [rpardini/docker-registry-proxy#27]
Hello,

did anyone tried using this with private Gitlab registry (and got it working)?

Thanks

**Comment by bcbrockway** [comment]
Did you manage to get this to work @panzelva? We're going to be looking for a caching solution for our GitLab Runners soon.

**Comment by panzelva** [comment]
I tried but with no luck :disappointed:

**Comment by rpardini** [comment]
Hi, I've released `0.3.0-beta1`, could you try with that version and report back? Thanks!

**Comment by panzelva** [comment]
@rpardini Hello, I tried adding
`-e AUTH_REGISTRIES="registry.gitlab.com:gitlab+deploy-token-156617:<some_password>"`
but with no luck - the docker pull ignores cache when trying to pull from GitLab registry.

I am using Deploy Token with `read_registry` rights.

```
docker run --rm --name docker_registry_proxy -it \
    -p 3128:3128 \
    -v $(pwd)/docker_mirror_cache:/docker_mirror_cache \
    -v $(pwd)/docker_mirror_certs:/ca \
    -e REGISTRIES="k8s.gcr.io gcr.io quay.io" \
    -e AUTH_REGISTRIES="registry.gitlab.com:gitlab+deploy-token-156617:<some_password>" \
    rpardini/docker-registry-proxy:0.3.0-beta2
```

Docker Hub and Quay both work flawlessly.

**Comment by rpardini** [comment]
Hmm, you need to add `registry.gitlab.com` to the `REGISTRIES` env var as well, eg,
```
-e REGISTRIES="k8s.gcr.io gcr.io quay.io registry.gitlab.com" \
-e AUTH_REGISTRIES="registry.gitlab.com:gitlab+deploy-token-156617:<some_password>" \
```
Otherwise it won't intercept those requests.

**Comment by betaboon** [comment]
@rpardini i just tried again against `registry.gitlab.com` using a deploy-token but i am still getting `401 Unauthorized`.
am i misinterpreting this issue being closed as being resolved?

this is what i ran:
```
docker run  --rm --name docker_registry_proxy -it \
  -p 3128:3128 \
  -v $(pwd)/docker_mirror_cache:/docker_mirror_cache \
  -v $(pwd)/docker_mirror_certs:/ca \
  -e REGISTRIES="registry.gitlab.com" \
  -e AUTH_REGISTRIES="registry.gitlab.com:USER:PASSWORD" \
  rpardini/docker-registry-proxy:0.3.0
```

am i missing something ?

**Comment by bephinix** [comment]
@betaboon Yes, you do. :wink: 

GitLab uses a different domain to handle the authentication procedure.
As Docker Hub uses `auth.docker.io` GitLab uses its primary (git) domain for the authentication.
If you run GitLab on `git.example.com` and its registry on `reg.example.com`, you need to include both in `REGISTRIES` and use the primary domain for `AUTH_REGISTRIES`.

```
docker run  --rm --name docker_registry_proxy -it \
  -p 3128:3128 \
  -v $(pwd)/docker_mirror_cache:/docker_mirror_cache \
  -v $(pwd)/docker_mirror_certs:/ca \
  -e REGISTRIES="reg.example.com git.example.com" \
  -e AUTH_REGISTRIES="git.example.com:USER:PASSWORD" \
  rpardini/docker-registry-proxy:0.3.0
```

For GitLab.com the authentication domain should be `gitlab.com`.

Maybe @rpardini can add this to the README. :+1:

**Comment by rpardini** [comment]
Thanks for this @bephinix -- I'll add this to the README right away.
@betaboon sorry I mismanaged your question, hope @bephinix 's answer helps you.

**Comment by betaboon** [comment]
@bephinix @rpardini thank you for these information. i will try this sometime soon and post back here if that solves my problem (as i suppose it does).

**Comment by AMarti96** [comment]
Hi all, I've been trying the solution introduced by @bephinix before, and I'm not able to make it work when the port of the private registry is mandatory.

The command I've been trying is the following:

```
docker run  --rm --name docker_registry_proxy -it \
-p 3128:3128  \
-v $(pwd)/docker_mirror_cache:/docker_mirror_cache  \
-v $(pwd)/docker_mirror_certs:/ca \
-e REGISTRIES="reg.example.com:4567 git.example.com" \
-e AUTH_REGISTRIES="git.example.com:<user>:<password>" \
rpardini/docker-registry-proxy:0.6.2
```

And also with similar variations including `reg.example.com:4567` in the AUTH_REGISTRIES part:

```
docker run  --rm --name docker_registry_proxy -it \
-p 3128:3128 \
-v $(pwd)/docker_mirror_cache:/docker_mirror_cache \
-v $(pwd)/docker_mirror_certs:/ca \
-e REGISTRIES="reg.example.com:4567 git.example.com" \
-e AUTH_REGISTRY_DELIMITER=":::" \
-e AUTH_REGISTRIES="reg.example.com:4567:::<user>:::<password> git.example.com:::<user>:::<password>" \
rpardini/docker-registry-proxy:0.6.2
```
But in both cases, I got the following response when pulling images from the GitLab private registry

```
~$ docker pull reg.example.com:4567/path/to/image:tag
Error response from daemon: Get https://reg.example.com:4567/v2/: Forbidden
```

Any idea/recommendation on how to proceed? Is there anything I may be missing here?

For more info, this is the internal logs from the `docker-registry-proxy` when running it in debug mode

<details>
<summary>Expand for logs related to the issue</summary>

<pre>

2021/04/06 21:10:55 [debug] 81#81: *1 accept: 192.168.1.100:55008 fd:17
2021/04/06 21:10:55 [debug] 81#81: *1 event timer add: 17: 60000:50260802
2021/04/06 21:10:55 [debug] 81#81: *1 reusable connection: 1
2021/04/06 21:10:55 [debug] 81#81: *1 epoll add event: fd:17 op:1 ev:80002001
2021/04/06 21:10:55 [debug] 81#81: timer delta: 13548
2021/04/06 21:10:55 [debug] 81#81: worker cycle
2021/04/06 21:10:55 [debug] 81#81: epoll timer: 60000
2021/04/06 21:10:55 [debug] 81#81: epoll: fd:17 ev:0001 d:00007FC7CE4E23C0
2021/04/06 21:10:55 [debug] 81#81: *1 http wait request handler
2021/04/06 21:10:55 [debug] 81#81: *1 malloc: 000055F8048483A0:1024
2021/04/06 21:10:55 [debug] 81#81: *1 recv: eof:0, avail:-1
2021/04/06 21:10:55 [debug] 81#81: *1 recv: fd:17 111 of 1024
2021/04/06 21:10:55 [debug] 81#81: *1 reusable connection: 0
2021/04/06 21:10:55 [debug] 81#81: *1 posix_memalign: 000055F8048B3000:4096 @16
2021/04/06 21:10:55 [debug] 81#81: *1 http process request line
2021/04/06 21:10:55 [debug] 81#81: *1 http request line: "CONNECT reg.example.com:4567 HTTP/1.1"
2021/04/06 21:10:55 [debug] 81#81: *1 http uri: ""
2021/04/06 21:10:55 [debug] 81#81: *1 http args: ""
2021/04/06 21:10:55 [debug] 81#81: *1 http exten: ""
2021/04/06 21:10:55 [debug] 81#81: *1 posix_memalign: 000055F8048B5000:4096 @16
2021/04/06 21:10:55 [debug] 81#81: *1 http process request header line
2021/04/06 21:10:55 [debug] 81#81: *1 http header: "Host: reg.example.com:4567"
2021/04/06 21:10:55 [debug] 81#81: *1 http header: "User-Agent: Go-http-client/1.1"
2021/04/06 21:10:55 [debug] 81#81: *1 http header done
2021/04/06 21:10:55 [debug] 81#81: *1 event timer del: 17: 50260802
2021/04/06 21:10:55 [debug] 81#81: *1 generic phase: 0
2021/04/06 21:10:55 [debug] 81#81: *1 generic phase: 1
2021/04/06 21:10:55 [debug] 81#81: *1 rewrite phase: 2
2021/04/06 21:10:55 [debug] 81#81: *1 http script value: "unknown-connect"
2021/04/06 21:10:55 [debug] 81#81: *1 http script set $docker_proxy_request_type
2021/04/06 21:10:55 [debug] 81#81: *1 rewrite phase: 4
2021/04/06 21:10:55 [debug] 81#81: *1 post rewrite phase: 5
2021/04/06 21:10:55 [debug] 81#81: *1 generic phase: 6
2021/04/06 21:10:55 [debug] 81#81: *1 generic phase: 7
2021/04/06 21:10:55 [debug] 81#81: *1 generic phase: 8
2021/04/06 21:10:55 [debug] 81#81: *1 access phase: 9
2021/04/06 21:10:55 [debug] 81#81: *1 access phase: 10
2021/04/06 21:10:55 [debug] 81#81: *1 access phase: 11
2021/04/06 21:10:55 [debug] 81#81: *1 post access phase: 12
2021/04/06 21:10:55 [debug] 81#81: *1 generic phase: 13
2021/04/06 21:10:55 [debug] 81#81: *1 generic phase: 14
2021/04/06 21:10:55 [debug] 81#81: *1 http finalize request: 403, "?" a:1, c:1
2021/04/06 21:10:55 [debug] 81#81: *1 http special response: 403, "?"
2021/04/06 21:10:55 [debug] 81#81: *1 http set discard body
2021/04/06 21:10:55 [debug] 81#81: *1 HTTP/1.1 403 Forbidden

</pre>
</details>

As you can see in there, it seems the remote registry answers with a 403, but the user/password are correct (I manually validated them to be sure they work in another environment and it was not a typo somewhere)

**Comment by khamadiev** [comment]
@AMarti96  Hi, I encountered the same problem with 403 status code, but it turns out this is proxy server responses. `proxy_connect` module by default allows connections on 443 and 563 ports. I've tried to change it by `proxy_connect_allow all;` in nginx.conf, but still no success, now docker client receives 302 redirect to gitlab login page :(

**Comment by AMarti96** [comment]
Many thanks for the pointer @khamadiev, it helped me to get it working in my environment.

### TL;DR

Take into account the following items when configuring this container to work with GitLab:

- If you run GitLab on `git.example.com` and its registry on `reg.example.com`, you need to include both in `REGISTRIES` and use the primary domain for `AUTH_REGISTRIES`.
- If you are running your registry on a specific port (`reg.example.com:4567`) in your GitLab instance, specify it with the port in `REGISTRIES` or you will get something like this when pulling:
```
HTTP 404 response body: unexpected end of JSON input: ""
```
- If you are using a backend different from the default one (like `s3`, `azure`, `gcs`...) be sure to add the backend domain to `REGISTRIES` and add a specific `location` in the nginx.conf file matching the path to the blobs and manifests.

### Long explanation

After fixing the `403 Forbidden` error by adding a `proxy_connect_allow all;` to the `/etc/nginx/nginx.conf` file, I was able to solve the 302 redirect by adding the main GitLab login page to the `AUTH_REGISTRIES` too.

As stated in previous comments, GitLab uses the main domain (`git.example.com` for example) for the auth part, even if you try to pull the images from a different registry url (`reg.example.com` for example).

With those changes, I managed to pull images and get logs in the docker container about the requests being processed, but **without caching**. After enabling debugging in the container and checking further the exact requests, I realized that the actual layers and manifests were not being pulled from the `reg.example.com` but from `s3.eu-central-1.amazonaws.com` (we were using S3 as a backend to store the images).

As a result, I ended up adding `s3.eu-central-1.amazonaws.com` to `REGISTRIES` too, but still no caching the layers when pulling them.

It seems that when getting the layers from S3, the objects are being pulled with this structure `^/<bucket>/docker/registry/v2/blobs/sha256/(.*)`, a structure that does not match what is being expected according to https://github.com/rpardini/docker-registry-proxy/blob/master/nginx.conf#L263, so I updated the `nginx.conf` file with the following:

```
        # For blob requests by digest, do cache, and treat redirects.
        location ~ ^/<bucket>/docker/registry/v2/blobs/sha256/(.*) {
            set $docker_proxy_request_type "blob-by-digest";
            include "/etc/nginx/nginx.manifest.common.conf";
        }
```

And I started to get HITS when pulling images! :smile: 

Some timing with a 6GB image to get the max out of this proxy:

```
# NO CACHE
$ time docker pull reg.example.io:4567/path/to/image:latest
real	2m19,866s
```
vs
```
# CACHE
$ time docker pull reg.example.io:4567/path/to/image:latest
real	0m20,969s
```

**Comment by bephinix** [comment]
@AMarti96 Great, that you figured it out! Maybe @rpardini can add this to the `README` for other users. :+1:

**Comment by hongkongkiwi** [comment]
This was very unclear for me and wasn't in the readme.... please add a note, for gitlab.com (which I'm sure many people are running, the registry and domain should be `gitlab.com` not git.gitlab.com or registry.gitlab.com

**Comment by rpardini** [comment]
Yes, I'll take PRs for README.md!

**Comment by Enissay** [comment]
> Many thanks for the pointer @khamadiev, it helped me to get it working in my environment.
> 
> ### TL;DR
> Take into account the following items when configuring this container to work with GitLab:
> 
> * If you run GitLab on `git.example.com` and its registry on `reg.example.com`, you need to include both in `REGISTRIES` and use the primary domain for `AUTH_REGISTRIES`.
> * If you are running your registry on a specific port (`reg.example.com:4567`) in your GitLab instance, specify it with the port in `REGISTRIES` or you will get something like this when pulling:
> 
> ```
> HTTP 404 response body: unexpected end of JSON input: ""
> ```
> 
> * If you are using a backend different from the default one (like `s3`, `azure`, `gcs`...) be sure to add the backend domain to `REGISTRIES` and add a specific `location` in the nginx.conf file matching the path to the blobs and manifests.
> 
> ### Long explanation
> After fixing the `403 Forbidden` error by adding a `proxy_connect_allow all;` to the `/etc/nginx/nginx.conf` file, I was able to solve the 302 redirect by adding the main GitLab login page to the `AUTH_REGISTRIES` too.
> 
> As stated in previous comments, GitLab uses the main domain (`git.example.com` for example) for the auth part, even if you try to pull the images from a different registry url (`reg.example.com` for example).
> 
> With those changes, I managed to pull images and get logs in the docker container about the requests being processed, but **without caching**. After enabling debugging in the container and checking further the exact requests, I realized that the actual layers and manifests were not being pulled from the `reg.example.com` but from `s3.eu-central-1.amazonaws.com` (we were using S3 as a backend to store the images).
> 
> As a result, I ended up adding `s3.eu-central-1.amazonaws.com` to `REGISTRIES` too, but still no caching the layers when pulling them.
> 
> It seems that when getting the layers from S3, the objects are being pulled with this structure `^/<bucket>/docker/registry/v2/blobs/sha256/(.*)`, a structure that does not match what is being expected according to https://github.com/rpardini/docker-registry-proxy/blob/master/nginx.conf#L263, so I updated the `nginx.conf` file with the following:
> 
> ```
>         # For blob requests by digest, do cache, and treat redirects.
>         location ~ ^/<bucket>/docker/registry/v2/blobs/sha256/(.*) {
>             set $docker_proxy_request_type "blob-by-digest";
>             include "/etc/nginx/nginx.manifest.common.conf";
>         }
> ```
> 
> And I started to get HITS when pulling images! smile
> 
> Some timing with a 6GB image to get the max out of this proxy:
> 
> ```
> # NO CACHE
> $ time docker pull reg.example.io:4567/path/to/image:latest
> real	2m19,866s
> ```
> 
> vs
> 
> ```
> # CACHE
> $ time docker pull reg.example.io:4567/path/to/image:latest
> real	0m20,969s
> ```

What permissions needed for the proxy access token on gitlab ?
